### PR TITLE
Implement monitoring interface deletion (rebased on dpdk-devel)

### PR DESCRIPTION
**Pull request by semihalf-berestovskyy-andriy**
- remove the RX queues mask (for some reason the commit significantly increases the latency)
- post-review fixes

- basic inter core communication
- ring queue releasing
- lock/unlock callbacks
- other minor changes

### CODE REVIEW

File: dpdk/vr_dpdk_interface.c [hunk 1]
```diff
@@ -356,12 +395,58 @@ dpdk_monitoring_if_add(struct vr_interface *vif)
...
+    monitored_vif = vrouter_get_interface(vif->vif_rid, monitored_vif_id);
+    if (!monitored_vif) {
+        RTE_LOG(ERR, VROUTER, "\terror getting vif to monitor: vif %u does not exist\n",
+                monitored_vif_id);
+        return -EINVAL;
+    }
+
+    /* stop monitoring */
+    dpdk_monitoring_stop(monitored_vif, vif);
+    vrouter_put_interface(monitored_vif);
```
Comment (srajag) [inline dpdk/vr_dpdk_interface.c]: Do we need to get and put in this function (i.e. can this just be __vrouter_get_interface)?

File: dpdk/vr_dpdk_interface.c [hunk 2]
```diff
@@ -317,13 +355,14 @@ dpdk_monitoring_if_add(struct vr_interface *vif)
...
     struct vr_interface *monitored_vif;
+    struct vrouter *router = vrouter_get(vif->vif_rid);
 
     RTE_LOG(INFO, VROUTER, "Adding monitoring vif %u KNI device %s"
                 " to monitor vif %u\n",
                 vif->vif_idx, vif->vif_name, monitored_vif_id);
 
     /* check if vif exist */
-    monitored_vif = vrouter_get_interface(vif->vif_rid, monitored_vif_id);
+    monitored_vif = __vrouter_get_interface(router, monitored_vif_id);
```
Comment (srajag) [inline dpdk/vr_dpdk_interface.c]: Don't we need a refcount on monitored vif? What happens if the monitored vif is deleted while it is being monitored?

File: dpdk/vr_dpdk_interface.c [hunk 3]
```diff
@@ -356,12 +395,58 @@ dpdk_monitoring_if_add(struct vr_interface *vif)
...
+                "vif %d KNI device does not exist\n",
+                vif->vif_idx);
+        return -EEXIST;
+    }
+
+    /* del the interface from the table of KNIs */
+    vr_dpdk.knis[vif->vif_idx] = NULL;
+
+    /* del the interface from the table of vHosts */
+    if (vr_dpdk.vhosts[vif->vif_idx]) {
```
Comment (srajag) [inline dpdk/vr_dpdk_interface.c]: Why does the monitoring interface need to be added to the table of vhosts?

File: dpdk/vr_dpdk_lcore.c [hunk 4]
```diff
@@ -266,6 +278,92 @@ vr_dpdk_lcore_if_schedule(struct vr_interface *vif, unsigned least_used_id,
...
+dpdk_lcore_rxtx_release(unsigned lcore_id, struct vr_interface *vif)
+{
+    struct vr_dpdk_lcore *lcore = vr_dpdk.lcores[lcore_id];
+    struct vr_dpdk_queue_params *rx_queue_params;
+    struct vr_dpdk_queue_params *tx_queue_params;
+
+    rx_queue_params = &lcore->lcore_rx_queue_params[vif->vif_idx];
+    if (rx_queue_params->qp_release_op) {
+        RTE_LOG(INFO, VROUTER, "\treleasing lcore %u RX queue\n", lcore_id);
+        rx_queue_params->qp_release_op(lcore_id, vif);
```
Comment (srajag) [inline dpdk/vr_dpdk_lcore.c]: Does vrouter_put_interface() get called on the q_vif for TX and RX?

File: dpdk/vr_dpdk_lcore.c [hunk 5]
```diff
@@ -363,76 +461,58 @@ vr_dpdk_packets_vroute(struct vr_interface *vif, struct vr_packet *pkts[VR_DPDK_
...
-    uint16_t num_tx_rings, max_tx_rings;
+    uint16_t nb_rtp;
 
-    total_pkts += dpdk_lcore_fwd_rx(lcore, &pkts[0], &rx_queues_mask);
-    total_pkts += dpdk_lcore_fwd_rx(lcore, &pkts[0], &rx_queues_mask);
-    total_pkts += dpdk_lcore_fwd_rx(lcore, &pkts[0], &rx_queues_mask);
-    total_pkts += dpdk_lcore_fwd_rx(lcore, &pkts[0], &rx_queues_mask);
-    total_pkts += dpdk_lcore_fwd_rx(lcore, &pkts[0], &rx_queues_mask);
+    total_pkts += dpdk_lcore_fwd_rx(lcore);
+    total_pkts += dpdk_lcore_fwd_rx(lcore);
```
Comment (srajag) [inline dpdk/vr_dpdk_lcore.c]: Why is this called twice?

File: dpdk/vr_dpdk_ringdev.c [hunk 6]
```diff
@@ -49,23 +57,93 @@ dpdk_ring_allocate(unsigned host_lcore_id, unsigned vif_idx, unsigned for_lcore_
...
+    rte_wmb();
+    lcore->lcore_nb_rings_to_push++;
+    RTE_VERIFY(lcore->lcore_nb_rings_to_push < VR_DPDK_MAX_RINGS);
+}
 
-    ring_to_push->rtp_tx_ring = tx_ring;
-    ring_to_push->rtp_tx_queue = tx_queue;
+/* Remove the ring from the list of rings to push */
+void
+dpdk_ring_to_push_remove(unsigned lcore_id, struct rte_ring *tx_ring)
```
Comment (srajag) [inline dpdk/vr_dpdk_ringdev.c]: This function will only be called from the netlink thread inside vrouter (i.e. there cannot be concurrent calls)?

File: dpdk/vr_dpdk_lcore.c [hunk 7]
```diff
@@ -82,55 +82,66 @@ vr_dpdk_lcore_least_used_get(void)
...
 
+/* Flush and remove TX queue from a lcore */
+static void
+dpdk_lcore_tx_queue_remove(struct vr_dpdk_lcore *lcore,
+                            struct vr_dpdk_queue *tx_queue)
+{
+    tx_queue->txq_ops.f_tx = NULL;
+    SLIST_REMOVE(&lcore->lcore_tx_head, tx_queue, vr_dpdk_queue,
+        q_next);
+    rte_wmb();
```
Comment (srajag) [inline dpdk/vr_dpdk_lcore.c]: Why wmb needed here - this is happening on the lcore that owns the tx queue (and no other lcore should be looking at its tx_queue list)?

File: dpdk/vr_dpdk_lcore.c [hunk 8]
```diff
@@ -82,55 +82,66 @@ vr_dpdk_lcore_least_used_get(void)
...
+dpdk_lcore_rx_queue_remove(struct vr_dpdk_lcore *lcore,
+                            struct vr_dpdk_queue *rx_queue)
+{
+    rx_queue->rxq_ops.f_rx = NULL;
+    SLIST_REMOVE(&lcore->lcore_rx_head, rx_queue, vr_dpdk_queue,
+        q_next);
+
+    /* decrease the number of RX queues */
+    lcore->lcore_nb_rx_queues--;
+    rte_wmb();
```
Comment (srajag) [inline dpdk/vr_dpdk_lcore.c]: Same as above - is wmb needed?

File: dpdk/vr_dpdk_lcore.c [hunk 9]
```diff
@@ -266,6 +278,92 @@ vr_dpdk_lcore_if_schedule(struct vr_interface *vif, unsigned least_used_id,
...
+    if (rx_queue_params->qp_release_op) {
+        RTE_LOG(INFO, VROUTER, "\treleasing lcore %u RX queue\n", lcore_id);
+        rx_queue_params->qp_release_op(lcore_id, vif);
+    }
+
+    tx_queue_params = &lcore->lcore_tx_queue_params[vif->vif_idx];
+    if (tx_queue_params->qp_release_op) {
+        RTE_LOG(INFO, VROUTER, "\treleasing lcore %u TX queue\n", lcore_id);
+        tx_queue_params->qp_release_op(lcore_id, vif);
+    }
```
Comment (srajag) [inline dpdk/vr_dpdk_lcore.c]: Would be good to set lcore_rx_queue_params to NULL (same for tx_params).

File: dpdk/vr_dpdk_interface.c (hunk 1 above)
Comment (semihalf-berestovskyy-andriy) [inline dpdk/vr_dpdk_interface.c]: Fixed.

File: dpdk/vr_dpdk_interface.c (hunk 2 above)
Comment (semihalf-berestovskyy-andriy) [inline dpdk/vr_dpdk_interface.c]: No, we don't need vif reference in order to monitor it. We use the VIF_FLAG_MONITORED to copy in/out packet to monitoring interface. If the monitored vif get deleted, we simply get no more packets. Then once the vifdump is interrupted, we delete the monitoring interface.

I added the comment.

File: dpdk/vr_dpdk_interface.c (hunk 3 above)
Comment (semihalf-berestovskyy-andriy) [inline dpdk/vr_dpdk_interface.c]: Fixed.

File: dpdk/vr_dpdk_lcore.c (hunk 5 above)
Comment (semihalf-berestovskyy-andriy) [inline dpdk/vr_dpdk_lcore.c]: RX operation for KNIs is quite expensive. We used rx_queue_mask to mask out the ports with no packets to read (i.e. read them less frequently). IMO we need to implement the same functionality for the list of RX queues now.

I added the TODO comment.

File: dpdk/vr_dpdk_ringdev.c (hunk 6 above)
Comment (semihalf-berestovskyy-andriy) [inline dpdk/vr_dpdk_ringdev.c]: Yes, it is called by the NetLink lcore only, but the list is used by the forwarding lcores. So I added some barriers to make sure we decrement the counter prior the queue removal.

File: dpdk/vr_dpdk_lcore.c (hunk 7 above)
Comment (semihalf-berestovskyy-andriy) [inline dpdk/vr_dpdk_lcore.c]: Removed.

File: dpdk/vr_dpdk_lcore.c (hunk 8 above)
Comment (semihalf-berestovskyy-andriy) [inline dpdk/vr_dpdk_lcore.c]: Removed.

**Pull request merged**